polyfill/lib/absolute.mjs

```diff
@@ -1,126 +1,197 @@
+    const value = GetSlot(this, EPOCHNANOSECONDS);
+    const epochMilliSeconds = value.ms;
+    return epochMilliSeconds;
+  }
+  getEpochMicroseconds() {
+    const value = GetSlot(this, EPOCHNANOSECONDS);
+    const epochNanoseconds = BigInt(value.ms) * BigInt(1e6) + (BigInt(value.ns) % BigInt(1e6));
+    return epochNanoseconds / BigInt(1e3);
+  }
+  getEpochNanoseconds() {
+    const value = GetSlot(this, EPOCHNANOSECONDS);
+    const epochMicroseconds = BigInt(value.ms) * BigInt(1e6) + (BigInt(value.ns) % BigInt(1e6));
+    return epochMicroseconds;
+  }
+
+  plus(durationLike = {}) {
+    const duration = ES.GetIntrinsic('%Temporal.duration%')(durationLike);
+    if (duration.years) throw new RangeError(`invalid duration field years`);
+    if (duration.months) throw new RangeError(`invalid duration field months`);
+
+    let { ms, ns } = GetSlot(this, EPOCHNANOSECONDS);
+    let negative = ms < 0 || ns < 0;
+    ns += duration.microseconds * 1000;
+    ns += duration.nanoseconds;
+    if (negative && ns > 0) {
+      ms += Math.floor(ns / 1e6);
+      ns = 1e6 - (ns % 1e6);
+    }
+    ms += duration.days * 86400000;
+    ms += duration.hours * 3600000;
+    ms += duration.minutes * 60000;
+    ms += duration.seconds * 1000;
+    ms += duration.milliseconds;
 
-Absolute.prototype.getEpochSeconds = function getEpochSeconds() {
-  let epochNanoSeconds = GetSlot(this, EPOCHNANOSECONDS);
-  let epochSecondsBigInt = epochNanoSeconds / 1000000000n;
-  let epochSeconds = ES.ToNumber(epochSecondsBigInt);
-  return epochSeconds;
-};
-Absolute.prototype.getEpochMilliseconds = function getEpochMilliseconds() {
-  let epochNanoSeconds = GetSlot(this, EPOCHNANOSECONDS);
-  let epochMillisecondsBigInt = epochNanoSeconds / 1000000n;
-  let epochMilliseconds = ES.ToNumber(epochMillisecondsBigInt);
-  return epochMilliseconds;
-};
-Absolute.prototype.getEpochMicroseconds = function getEpochMicroseconds() {
-  let epochNanoSeconds = GetSlot(this, EPOCHNANOSECONDS);
-  let epochMicroseconds = epochNanoSeconds / 1000n;
-  return epochMicroseconds;
-};
-Absolute.prototype.getEpochNanoseconds = function getEpochNanoseconds() {
-  let epochNanoSeconds = GetSlot(this, EPOCHNANOSECONDS);
-  return epochNanoSeconds;
-};
+    if (negative && ms > 0) {
+      ms -= 1;
+      ns += 1e6;
+    }
 
-Absolute.prototype.plus = function plus(durationLike = {}) {
-  const Duration = ES.GetIntrinsic('%Temporal.Duration%');
-  const duration = new Duration(durationLike);
-  if (duration.years) throw new RangeError(`invalid duration field years`);
-  if (duration.months) throw new RangeError(`invalid duration field months`);
+    const result = Object.create(Absolute.prototype);
+    CreateSlots(result);
+    SetSlot(result, EPOCHNANOSECONDS, { ms, ns });
+    return result;
+  }
+  minus(durationLike = {}) {
+    const duration = ES.GetIntrinsic('%Temporal.duration%')(durationLike);
+    if (duration.years) throw new RangeError(`invalid duration field years`);
```

Comment:
I think it's important not to use falsiness here. I would suggest comparing to undefined instead.

Reply:
Ok, but after the cast the value is guaranteed to be a number >= 0

Reply:
0 is falsy